crates/turbo-tasks/src/vc/cast.rs

```diff
@@ -26,7 +26,20 @@ where
     type Output = ReadRef<T>;
 
     fn cast(content: CellContent) -> Result<Self::Output> {
-        content.cast::<T>()
+        Ok(
+            // Safety: the `VcValueType` implementor must guarantee that both `T` and
+            // `Repr` are #[repr(transparent)].
+            unsafe {
+                // Downcast the cell content to the expected representation type, then
+                // transmute it to the expected type.
+                std::mem::transmute_copy::<
+                    ManuallyDrop<ReadRef<<T::Read as VcRead<T>>::Repr>>,
+                    Self::Output,
+                >(&ManuallyDrop::new(
+                    content.cast::<<T::Read as VcRead<T>>::Repr>()?,
+                ))
```

Comment:
The `ManuallyDrop` are here because `std::mem::transmute` doesn't support generic types, and [this message](https://users.rust-lang.org/t/transmute-doesnt-work-on-generic-types/87272/9) recommended this option instead.

I've added a comment to refer to the message in the code.